### PR TITLE
[RFC] Enable syntax and filetypes by default

### PR DESCRIPTION
**Pull request by fmoralesc**
Picks up on #3767 

I also found that clint.py will find some false positives for lines beggining in "for", so I made a small tweak to it (would appreciate comments on that).

### CODE REVIEW

**Comment by jamessan** [comment]
I've advocated against a change like this in various forums in the past.  My reasoning is that it makes behavior less consistent for the novice user by doing this for them instead of letting them do it in their vimrc at the appropriate time.

If a user is modifying `'runtimepath'`, either on their own or through a plugin manager, then enabling filetype before they've done so excludes any potential ftdetect scripts they have in the new runtime directories from being sourced.  They have to know to `filetype off` before modifying `'runtimepath'` and then turn it on again afterwards.

I think most plugin managers handle this now, but why put more steps in front of a novice user?

**Comment by justinmk** [comment]
Hmm, we didn't think of that, I think.

**Comment by bfredl** [comment]
I don't know the exact logic here but would it be completely unreasonable to fix the underlying problem, by say, sourcing new ftdetect scripts when `&rtp` changes?

**Comment by fmoralesc** [comment]
> If a user is modifying 'runtimepath', either on their own or through a plugin manager, then enabling filetype before they've done so excludes any potential ftdetect scripts they have in the new runtime directories from being sourced.

This is the reason this is done at this point in the initialization, so those conflicts don't arise. The syntax and filetypes are enabled *after* the vimrc is sourced, but *before* loading plugins, so the user can easily override the runtime. I think this is a non-issue.

Also, I think this is something plugin managers should *not* handle, but in that I guess I'm just contrarian. 

My question here is : *Why* ask the user to enable this functionality themselves? I think it's just silly to have nvim not enable this by default.

**Comment by jamessan** [comment]
> This is the reason this is done at this point in the initialization, so those conflicts don't arise. The syntax and filetypes are enabled after the vimrc is sourced, but before loading plugins, so the user can easily override the runtime.

Ok.  That wasn't clear to me.  If that's the case, then I don't have any objections.

> Also, I think this is something plugin managers should not handle, but in that I guess I'm just contrarian.

No, I agree here.  However, they have to do this with Vim, otherwise the plugins being managed won't necessarily behave as expected, due to exactly the reasons I described. :) Since you've fixed the timing issue, then you're right that plugin managers won't need to do the filetype dance with Neovim.

**Comment by fmoralesc** [comment]
Glad to see we agree ;) 

That is also the reason we do this in C instead of in a viml script in the runtime, execution at this point in the initialization is not guaranteed.

**Comment by justinmk** [comment]
Great! Glad we got some feedback on this finally.

If there are no other objections let's put this in 0.1.3.

**Comment by fmoralesc** [comment]
:+1: 

I still want some feedback on the change I made on clint.py, I think it is right but I don't want to screw it up for others. Perhaps change my '[ (]' to '[\s(]'?

File: src/nvim/ex_docmd.c [hunk 1]
```diff
@@ -9293,6 +9293,24 @@ static void ex_filetype(exarg_T *eap)
     EMSG2(_(e_invarg2), arg);
 }
 
+/// Same as :filetype plugin indent enable
```
Comment (jamessan) [inline src/nvim/ex_docmd.c]: `on`, not `enable`.  `enable` is just for the `:syntax` command.  The same typo is in main.c.

**Comment by jamessan** [comment]
So, how does this interact with a user specifically calling `filetype plugin on` in their init.vim?  Some people may only want ftplugins enabled, not filetype indenting.

**Comment by bfredl** [comment]
personally I would have added a requirement that `for|while|if` isn't followed by an alphanumeric_ but I think your soulution should work as well... Have you tried that it hasn't any obvious false negative?

File: src/nvim/ex_docmd.c (hunk 1 above)
Comment (fmoralesc) [inline src/nvim/ex_docmd.c]: :+1:

**Comment by fmoralesc** [comment]
> So, how does this interact with a user specifically calling filetype plugin on in their init.vim? Some people may only want ftplugins enabled, not filetype indenting.
 
@jamessan That case, I didn't think about. It currently overrides it. A test can be done so if it finds that any of the flags are already set, it won't force the enabling of the runtime.

**Comment by fmoralesc** [comment]
@bfredl I think this covers all the obvious cases, but it might perhaps fail if there is a tab after the keyword, or something like that, which is why I though perhaps `'[\s(]'` could work better...

File: src/nvim/main.c [hunk 2]
```diff
@@ -332,6 +332,12 @@ int main(int argc, char **argv)
   /* Source startup scripts. */
   source_startup_scripts(&params);
 
+  // If using the runtime (-u is not NONE), enable syntax and filetype plugins
+  if (params.use_vimrc != NULL && strcmp(params.use_vimrc, "NONE") != 0) {
+    syn_cmd("syntax");        // Enable syntax highlighting.
```
Comment (jamessan) [inline src/nvim/main.c]: `force_enable_filetype` should come first, because when it loads `syntax/syntax.vim` that will cause `filetype on` to run.

Comment (fmoralesc) [inline src/nvim/main.c]: Thanks for pointing that out :+1:

**Comment by justinmk** [comment]
squashed and made some documentation changes (e.g. the startup order needed an update) here: https://github.com/neovim/neovim/pull/4252

I noticed that there doesn't seem to be a good way to totally disable these defaults from the user's vimrc. E.g. `nvim --cmd 'filetype off plugin off indent off'` causes the exact opposite of the user's intentions. I imagine this might be viewed as a bug...

**Comment by fmoralesc** [comment]
Hm. We should add a flag in the initialization process to check whether the filetypes are explicitely disabled (or make the currents flags integers, so, for example, -1 means not set, 0 means disabled and 1 means set).

**Comment by jamessan** [comment]
The current flags are integers.

**Comment by fmoralesc** [comment]
For a moment I was under the impression they were bools, but you are totally right, @jamessan. Then it's easily fixable.

Are we closing this PR, @justinmk?

**Comment by justinmk** [comment]
@fmoralesc if you want to change the flags to an enum, can you update this PR from mine?

**Comment by fmoralesc** [comment]
Done in #4255 (I messed up my local branch).

**Pull request closed without merging**